### PR TITLE
Component/4080-badge

### PR DESCRIPTION
**Pull request by brianacnguyen**
**Description**

- Componentize Badge Component
- Add Badge capability to Avatar Base

**Screenshots/Recordings**
Simulator Screen Shot - iPhone 11 Pro - 2022-08-11 at 20 52 59

**Issue**

Progresses #???

**Checklist**

* [ ] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

### CODE REVIEW

File: app/component-library/components/Avatars/AvatarBase/AvatarBase.styles.ts [hunk 1]
```diff
@@ -13,9 +16,39 @@ import { AvatarBaseStyleSheetVars } from './AvatarBase.types';
  */
 const styleSheet = (params: { vars: AvatarBaseStyleSheetVars }) => {
   const {
-    vars: { style, size },
+    vars: { style, size, badgePosition },
   } = params;
   const sizeAsNum = Number(size);
+  let position;
```
Comment (Cal-L) [inline app/component-library/components/Avatars/AvatarBase/AvatarBase.styles.ts:22]: Can omit the autos by spreading this position in the style below

File: app/component-library/components/Avatars/AvatarBase/AvatarBase.styles.ts [hunk 2]
```diff
@@ -24,9 +57,20 @@ const styleSheet = (params: { vars: AvatarBaseStyleSheetVars }) => {
         width: sizeAsNum,
         borderRadius: sizeAsNum / 2,
         overflow: 'hidden',
+        backgroundColor: 'white',
```
Comment (Cal-L) [inline app/component-library/components/Avatars/AvatarBase/AvatarBase.styles.ts:60]: background.default from theme colors

File: app/component-library/components/Avatars/AvatarBase/AvatarBase.types.ts [hunk 3]
```diff
@@ -1,6 +1,5 @@
 // Third party dependencies.
 import { ViewProps } from 'react-native';
-
```
Comment (Cal-L) [inline app/component-library/components/Avatars/AvatarBase/AvatarBase.types.ts:3]: undo

File: app/component-library/components/Avatars/AvatarBase/AvatarBase.styles.ts [hunk 4]
```diff
@@ -24,9 +57,20 @@ const styleSheet = (params: { vars: AvatarBaseStyleSheetVars }) => {
...
       } as ViewStyle,
       style,
     ) as ViewStyle,
+    badgeContent: {
+      top: position.top,
+      right: position.right,
+      bottom: position.bottom,
+      left: position.left,
+      width: 16,
```
Comment (Cal-L) [inline app/component-library/components/Avatars/AvatarBase/AvatarBase.styles.ts:69]: Maybe use AvatarBaseSize.Xs and convert to number to use here

File: app/component-library/components/Avatars/AvatarBase/AvatarBase.styles.ts [hunk 5]
```diff
@@ -24,9 +57,20 @@ const styleSheet = (params: { vars: AvatarBaseStyleSheetVars }) => {
...
+      height: 16,
+      borderRadius: 8,
```
Comment (Cal-L) [inline app/component-library/components/Avatars/AvatarBase/AvatarBase.styles.ts:71]: same here

File: app/component-library/components/Badge/README.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
+# Badge
+
+Badge is a wrapper component that enables any custom component to be attached to a child component.
+
+## Props
+
+This component extends `ViewProps` from React Native's [View Component](https://reactnative.dev/docs/view).
```
Comment (Cal-L) [inline app/component-library/components/Badge/README.md:7]: Pull Component out after anchor

**Review by Cal-L — APPROVED** [review]
Left some comments - LGTM after!

File: app/component-library/components/Badge/Badge.tsx [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  style,
+  ...props
+}) => {
+  const { styles } = useStyles(styleSheet, {
+    style,
+    badgeContentStyle,
+  });
+
+  return (
+    <View style={styles.base} {...props}>
```
[the remark this answers is not in the document]
Comment (brianacnguyen) [reply app/component-library/components/Badge/Badge.tsx:27]: After doing a little digging, React.Fragment would not work here as the badge needs an actual rendered element to attach the position to. will keep as is

File: app/component-library/components/Badge/Badge.styles.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ * @returns StyleSheet object.
+ */
+const styleSheet = (params: { theme: Theme; vars: BadgeStyleSheetVars }) => {
+  const { vars } = params;
+  const { style, badgeContentStyle } = vars;
+
+  return StyleSheet.create({
+    base: Object.assign(
+      {
+        alignSelf: 'flex-start',
```
[the remark this answers is not in the document]
Comment (brianacnguyen) [reply app/component-library/components/Badge/Badge.styles.ts:25]: answer below

**Comment by jpcloureiro** [comment]
I see that only the AvatarBase has the ability to position the badge. Is BadgeWrapper not allowed to position the badge other than on Top Right corner?

**Comment by brianacnguyen** [comment]
> 

As of right now, the positioning preset for the badge can only apply to avatar badges. In the design spec, the positioning of top: -4 and right: -4 for TopRight and bottom: -4 and right: -4, can ONLY apply to an XS size avatar. If you place a Tag as a badge with badgeWrapper with the preset positioning, it does not look good and requires different positioning. As such, I've decided to keep the current positioning logic with the AvatarBadge until the designers can fully figure out a positioning logic/formula that can be applied to ANY badge.

**Review by jpcloureiro — APPROVED** [review]
Thanks for the explanation @brianacnguyen

Looks good to me! :shipit:

File: app/component-library/components/Avatars/AvatarBase/AvatarBase.styles.ts [hunk 9]
```diff
@@ -1,8 +1,15 @@
...
 
+// External dependencies.
+import { Theme } from '../../../../util/theme/models';
+
 // Internal dependencies.
-import { AvatarBaseStyleSheetVars } from './AvatarBase.types';
+import {
+  // AvatarBaseSize,
+  AvatarBaseStyleSheetVars,
+  // AvatarBadgePositionVariant,
```
Comment (Cal-L) [inline app/component-library/components/Avatars/AvatarBase/AvatarBase.styles.ts:11]: Remove commented out code

File: app/component-library/components/Avatars/AvatarBase/README.md [hunk 10]
```diff
@@ -14,15 +14,35 @@ Optional enum to select between size variants.
 | :-------------------------------------------------- | :------------------------------------------------------ | :----------------------------------------------------- |
 | [AvatarBaseSize](./AvatarBase.types.ts#L6)          | Yes                                                     | Md                                                     |
 
+### `badgeContent`
```
Comment (Cal-L) [inline app/component-library/components/Avatars/AvatarBase/README.md:17]: Remove badge related stuff

File: app/component-library/components/Badges/Badge/variants/BadgeNetwork/README.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+This component extends `ViewProps` from React Native's [View](https://reactnative.dev/docs/view) component.
+
+### `variant`
+
+Variant of badge.
+
+| <span style="color:gray;font-size:14px">TYPE</span> | <span style="color:gray;font-size:14px">REQUIRED</span> |
+| :-------------------------------------------------- | :------------------------------------------------------ |
+| [BadgeVariant.Network](../../Badge.types.ts#L7)                                           | Yes                                                     |
```
Comment (Cal-L) [inline app/component-library/components/Badges/Badge/variants/BadgeNetwork/README.md:15]: BadgeVariants.Network

**Review by Cal-L — APPROVED** [review]
1 or 2 comments - LGTM!

**Pull request merged**